log: improve checkaddrman logging with duration in milliseconds

jonatack opened the pull request:
This patch:
- updates the `logging/timer.h::Timer` class to allow not repeating the log message on completion
- adds a `LOG_TIME_MILLIS_WITH_CATEGORY_MSG_ONCE` macro that prints the descriptive message when logging the start but not when logging the completion
- updates the checkaddrman logging to log the duration, and renames the function like the `-checkaddrman` configuration option in order to prefix every log message with `CheckAddrman` instead of the longer, less pleasant, and different-from-checkaddrman `ForceCheckAddrman` (the Doxygen documentation on the function already makes clear that it is unaffected by `m_consistency_check_ratio`).

before
```
2021-09-21T18:42:50Z [opencon] Addrman checks started: new 64864, tried 1690, total 66554
2021-09-21T18:42:50Z [opencon] Addrman checks completed successfully
```

after
```
2021-09-21T18:42:50Z [opencon] CheckAddrman: new 64864, tried 1690, total 66554 started
2021-09-21T18:42:50Z [opencon] CheckAddrman: completed (76.21ms)
```

To test, build and run bitcoind with `-debug=addrman -checkaddrman=<n>` for a value of `n` in the range of, say, 10 to 40.

jonatack committed:
p2p: improve checkaddrman logging with duration in milliseconds

and update the function name to CheckAddrman (drop "Force") for
nicer log output as it is prefixed to each of these log messages:

2021-09-21T18:42:50Z [opencon] CheckAddrman: new 64864, tried 1690, total 66554 started
2021-09-21T18:42:50Z [opencon] CheckAddrman: completed (76.21ms)

The existing Doxygen documentation on the function already makes
clear that it is unaffected by m_consistency_check_ratio.

MarcoFalke commented on src/addrman.h:
```diff
@@ -404,15 +404,15 @@ class CAddrMan
 
     //! Perform consistency check. Returns an error code or zero.
-    int Check_() const EXCLUSIVE_LOCKS_REQUIRED(cs);
+    int CheckAddrMan() const EXCLUSIVE_LOCKS_REQUIRED(cs);
```
any reason for the rename here? This will just cause conflicts and increase the git history without any benefit.

jonatack replied:
It's to have the log print "CheckAddrMan:" instead of "Check_:"

laanwj replied:
All the function names seem to end with `_`, might want to keep that at least for consistency?

jonatack replied:
Done, updated the pull description

practicalswift commented:
Concept ACK

theStack commented:
Concept ACK

laanwj commented:
Concept ACK

MarcoFalke commented:
cr-only ACK d01a6c77ce

MarcoFalke commented on src/addrman.cpp:
```diff
@@ -748,15 +749,16 @@ CAddrInfo CAddrMan::Select_(bool newOnly) const
 
-    LogPrint(BCLog::ADDRMAN, "Addrman checks started: new %i, tried %i, total %u\n", nNew, nTried, vRandom.size());
+    LOG_TIME_MILLIS_WITH_CATEGORY(
+        strprintf("new %i, tried %i, total %u", nNew, nTried, vRandom.size()), BCLog::ADDRMAN);
```
Looks like this will print the statistics twice, implying they would change in-between, which might be confusing.

Not sure how to best solve this.

jonatack replied:
> Looks like this will print the statistics twice, implying they would change in-between, which might be confusing.
> 
> Not sure how to best solve this.

I'll have a look.

jonatack replied:
Done, example on signet:
```
2021-09-11T19:39:05Z [scheduler] Enter: lock contention ::cs_main, validation.cpp:4874 started
2021-09-11T19:39:05Z [opencon] CheckAddrMan_: new 442, tried 57, total 499 started
2021-09-11T19:39:05Z [opencon] CheckAddrMan_: completed (10.30ms)
2021-09-11T19:39:07Z [scheduler] Enter: lock contention ::cs_main, validation.cpp:4874 completed (2906703μs)
```

jonatack committed:
log, timer: allow not repeating log message on completion

jonatack committed:
log, timer: add LOG_TIME_MILLIS_WITH_CATEGORY_MSG_ONCE macro

that prints the descriptive message when logging the start
but not when logging the completion.

jonatack commented:
Rebased and updated following the merge of #22734.

jonatack commented:
Rebased and updated following the merges of #22734 and #22831.

MarcoFalke commented on src/logging/timer.h:
```diff
@@ -91,6 +100,8 @@ class Timer
 #define LOG_TIME_MILLIS_WITH_CATEGORY(end_msg, log_category) \
     BCLog::Timer<std::chrono::milliseconds> PASTE2(logging_timer, __COUNTER__)(__func__, end_msg, log_category)
+#define LOG_TIME_MILLIS_WITH_CATEGORY_MSG_ONCE(end_msg, log_category) \
+    BCLog::Timer<std::chrono::milliseconds> PASTE2(logging_timer, __COUNTER__)(__func__, end_msg, log_category, /* msg_on_completion=*/false)
```
unrelated idea:

The following log could also be improved with this new macro:

```
src/validation.cpp:                LogPrintf("[snapshot] flushing coins cache (%.2f MB)... ", /* Continued */

jonatack replied:
Thanks Marco, will have a look and comment here.

MarcoFalke replied:
@jonatack

jonatack replied:
Oh, thanks for the ping!

jonatack replied:
#23738

jonatack commented:
Rebased.

jonatack commented:
Rebased!

laanwj commented:
Code review ACK 22b44fc696dc1078c40d17e2d497c74c7b4ae750

MarcoFalke commented:
Merged
